08-rails/how-to-use-postgres.md

```diff
@@ -0,0 +1,124 @@
+
+# Setting Up A Rails Project with Postgres
+
+Postgres (or PostgreSQL) is an open-source database which can provide scalability to your application.  By default Rails development sites use sqlite3 which can work fine, but it can be easier to use the same database application in development & production.  
+
+## SQLite vs Postgres
+
+SQLite is a "serverless" database which means there isn't a separate database program running to manage access to the data, instead the application accesses an SQLite file directly, similar to how our previous Ruby programs accessed CSV files.  
+
+The advantage is that there's almost no configuration involved in setting up an SQLite database and no access-control setup.  So SQLite is simple to get up and running.  
+
+SQLite runs into trouble when multiple programs access the same database directly.  There's not a server coordinating access and keeping two users from writing to the same data.  Instead SQLite uses file blocks to keep one program from accessing the data while another is writing to it.  This is far less efficient than using a server to coordinate access.  So an SQLite database is not well suited to a site with lots of simultaneous accesses.  It can however serve for sites which receive low levels of traffic and which do not share a database with other applications.  For the most-part we do not use SQLite for a production website.  However it is a very easy-to-use database for development.
+
+Postgres is a more traditional client-server database which runs a separate process that Rails will communicate with to access data.  
+
+The downside is that we need to setup a Postgres server and deal with some configuration of the database before we can use it.  
+
+On the upside using Postgres on both the development and production ends means reduces any chances of problems when deploying our app on the web via AWS or Heroku. 
+
+
+## Installing Postgres
+
+To install Postgres you can use the Brew package manager, then set brew to start postgres to launch every time the machine reboots and lastly install the pg gem to allow Ruby to talk to the database.  
+
+These steps you only need perform once.
+
+```bash
+$ brew install postgresql
+...
+$ brew services start postgresql
+...
+$ gem install pg
+...
+```
+
+## To setup a rails project with Postgres
+
+After Postgres is installed and configured you can create new rails apps with postgres this way.  `rails db:migrate` is an essential step because it sets up the required tables in the database for a successful start of the rails server.  
+
+```bash
+$ rails new my-rails-project -d postgresql
```

Comment:
If we instruct them to create a `.railsrc` file in their home directory with `-d postgresql` in it, they won't need to add it every time they use `rails new`:

```bash
echo "-d postgresql" > ~/.railsrc
```